core/client/fs/src/main/java/alluxio/client/block/stream/DefaultBlockWorkerClient.java

```diff
@@ -73,182 +72,219 @@ public class DefaultBlockWorkerClient implements BlockWorkerClient {
-        mRpcChannel = GrpcChannelBuilder.newBuilder(address, alluxioConf)
-            .setSubject(userState.getSubject())
-            .setNetworkGroup(GrpcNetworkGroup.RPC)
-            .build();
-        lastException = null;
-        break;
-      } catch (StatusRuntimeException e) {
-        close();
-        throw AlluxioStatusException.fromStatusRuntimeException(e);
-      } catch (UnauthenticatedException e) {
-        close();
-        userState.relogin();
-        lastException = e;
-      }
-    }
-    if (lastException != null) {
-      throw lastException;
-    }
-    mStreamingAsyncStub = BlockWorkerGrpc.newStub(mStreamingChannel);
-    mRpcBlockingStub = BlockWorkerGrpc.newBlockingStub(mRpcChannel);
-    mRpcAsyncStub = BlockWorkerGrpc.newStub(mRpcChannel);
+  public DefaultBlockWorkerClient(ClientContext context, GrpcServerAddress address,
+                                  AlluxioConfiguration alluxioConf) {
+    // BlockWorkerClient is typically used inside a BlockInStream
+    // to fetch block chunk. BlockInStream has its own retry mechanism, so we don't
+    // retry at this level so as not to interfere with the retry of our higher-level
+    // callers.
+    // An instance where retries interfere with each other is when we and our calling
+    // BlockInStream both adopted a timer-based retry, then all the time could be wasted
+    // in our inner retry, and the outer stream has no chance of retrying a different client.
+    super(context, null, () -> new CountingRetry(0));
+    // the server address this client connects to. It might not be an IP address
     mAddress = address;
     mRpcTimeoutMs = alluxioConf.getMs(PropertyKey.USER_RPC_RETRY_MAX_DURATION);
     mTracker = DETECTOR.track(this);
   }
 
+  @Override
+  protected void beforeConnect() {}
+
+  @Override
+  protected void afterConnect() {
+    // these are alias to the channel built in parent class
+    mStreamingChannel = Preconditions.checkNotNull(
+            mAdditionalChannels.get(GrpcNetworkGroup.STREAMING));
+    mRpcChannel = Preconditions.checkNotNull(mChannel);
+
+    // set up stubs
+    mStreamingAsyncStub = BlockWorkerGrpc.newStub(mStreamingChannel);
+    mRpcAsyncStub = BlockWorkerGrpc.newStub(mRpcChannel);
+    mRpcBlockingStub = BlockWorkerGrpc.newBlockingStub(mRpcChannel);
+  }
+
   @Override
   public boolean isShutdown() {
-    return mStreamingChannel.isShutdown() || mRpcChannel.isShutdown();
+    return isClosed();
   }
 
   @Override
   public boolean isHealthy() {
-    return !isShutdown() && mStreamingChannel.isHealthy() && mRpcChannel.isHealthy();
+    if (isShutdown()) {
+      // client is shut down
+      return false;
+    }
+
+    if (isConnected()) {
+      return mStreamingChannel != null && mStreamingChannel.isHealthy()
+              && mRpcChannel != null && mRpcChannel.isHealthy();
+    }
+
+    // client is not connected yet, assume it is healthy
+    return true;
   }
```

Comment:
I'm a little bit uncomfortable about this check. It's subject to the risk of false negative: it might falsely think some unhealthy client is healthy because `retryRPC` disconnects on rpc failure, and hence a client that has failed several rpc calls is indistinguishable from a fresh client that has not connected yet.